### DISCUSSION TITLE
"require_motion" --> is my context correct?

### DISCUSSION THREAD

**Discussion by gniknalu** [roflcoopter/viseron#186] [Q&A] [answered]
Do I have "require_motion" in the correct context?? 

```
cameras:
  - name: "FrontWest"
    host: !secret xxxxxxxxxxxxxxxxx
    port: !secret xxxxxxxxxxxxxxxxx
    stream_format: rtsp
    username: !secret xxxxxxxxxxxxxxxxx
    password: !secret xxxxxxxxxxxxxxxxx
    path: xxxxxxxxxxxxxxxxxxxxxxxxxx
    mqtt_name: xxxxxxxxxxxxxxxxx
    publish_image: true
    zones:
      - name: DrivewayRegion
        points:
          - x: 213
            y: 158
            <............>
        labels:
          - label: person
            confidence: 0.8
            triggers_recording: true
          - label: truck
            confidence: 0.8
            triggers_recording: true
            require_motion: true
      - name: noDetectZone
        points:
          - x: 410
            y: 0
            <............>
        labels:
          - label: dog
            confidence: 1
            triggers_recording: false
```

I didn't want it to constantly report/record **my** truck. 

Where I added "require_motion: true"  generates this error:

```
voluptuous.error.MultipleInvalid: extra keys not allowed @ data['cameras'][0]['zones'][0]['labels'][1]['require_motion']
Traceback (most recent call last):
  File "viseron.py", line 8, in <module>
    from lib.config import CONFIG, NVRConfig, ViseronConfig
  File "/src/viseron/lib/config/__init__.py", line 101, in <module>
    CONFIG = VISERON_CONFIG_SCHEMA(raw_config)
  File "/usr/local/lib/python3.6/dist-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/usr/local/lib/python3.6/dist-packages/voluptuous/schema_builder.py", line 594, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/usr/local/lib/python3.6/dist-packages/voluptuous/schema_builder.py", line 432, in validate_mapping
    raise er.MultipleInvalid(errors)
```

**Comment by roflcoopter** [comment]
Hmm it seems correct. What does your docker command look like? Im thinking you might be running an older version

**Reply by gniknalu** [reply]
This is my second time working with viseron - this one was downloaded/installed just this week. This is my command:

`docker run -d --restart unless-stopped -v /usr/share/viseron/media:/recordings -v /usr/share/viseron:/config -v /etc/localtime:/etc/localtime:ro --name viseron --runtime=nvidia docker.io/roflcoopter/viseron-cuda:latest`

**Reply by roflcoopter** [reply]
Ahh i see yes that is the old image name. Newer one is called roflcoopter/amd64-cuda-viseron

**Reply by gniknalu** [reply]
I **DID** try and pull amd64-cuda-viseron originally but got this error:

```
docker pull roflcoopter/amd64-cuda-viseron
Using default tag: latest
Error response from daemon: manifest for roflcoopter/amd64-cuda-viseron:latest not found: manifest unknown: manifest unknown
```

But didn't get the error when using the "docker pull roflcoopter/viseron-cuda" command.

This is my docker-compose

```
version: "2.4"

services:
  viseron:
    image: roflcoopter/amd64-cuda-viseron:latest
    container_name: viseron
    volumes:
      - /usr/share/viseron/media/:/recordings
      - /usr/share/viseron/:/config
      - /etc/localtime:/etc/localtime:ro
    runtime: nvidia

```

I've been documenting my journey - hence I have a file of everything I tried.

**Reply by roflcoopter** [reply]
Interesting! For some reason the tag `latest` has not been pushed. Need to check my CI/CD pipelines.
In the meantime you can pull roflcoopter/amd64-cuda-viseron:1.8.0 to get the latest version.

**Reply by roflcoopter** [reply]
I manually pushed the latest tag. On next release the scripts will push it properly automatically

**Reply by gniknalu** [reply]
pull roflcoopter/amd64-cuda-viseron:1.8.0 is working!

I'll let you know what happens when i add "require_motion" back

**Thank you!**

**Reply by gniknalu** [reply]
Still seeing the same issue with version 1.8.0
Here is a redacted portion of my config where I tested it:

```
  - name: xxxxxxxxxxxx
    host: !secret xxxxxxxxxxxx
    port: !secret xxxxxxxxxxxx
    stream_format: rtsp
    username: !secret xxxxxxxxxxxx
    password: !secret xxxxxxxxxxxx
    path: /xxxxxxxxxx
    mqtt_name: xxxxxxxxxxxx
    publish_image: true
    zones:
      - name: xxxxxxxxxxxx
        points:
          - x: 670
            y: 1073
         <..........>
        labels:
          - label: person
            confidence: 0.8
            triggers_recording: true
          - label: car
            confidence: 0.8
            triggers_recording: true
          - label: truck
            confidence: 0.8
            triggers_recording: true
            require_motion: true
```

And this is what I received in the log:

```
Traceback (most recent call last):
  File "viseron.py", line 8, in <module>
    from lib.config import CONFIG, NVRConfig, ViseronConfig
  File "/src/viseron/lib/config/__init__.py", line 101, in <module>
    CONFIG = VISERON_CONFIG_SCHEMA(raw_config)
  File "/usr/local/lib/python3.6/dist-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/usr/local/lib/python3.6/dist-packages/voluptuous/schema_builder.py", line 594, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/usr/local/lib/python3.6/dist-packages/voluptuous/schema_builder.py", line 432, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: extra keys not allowed @ data['cameras'][1]['zones'][0]['labels'][4]['require_motion']
```

**Reply by roflcoopter** [reply]
You sure you recreated the container properly? It works fine for me.

**Reply by gniknalu** [reply]
You're right - I missed a step. Re-doing.

**Reply by gniknalu** [reply]
It's the strangest thing. It **appears** to start up with the setting fine.  But I'm **guessing** when it detects "truck" it errors out?

This is what I'm seeing in the log:

```
[2021-05-28 11:03:54] [viseron.recorder.fronteast] [DEBUG   ] - Initializing ffmpeg recorder
[2021-05-28 11:03:55] [viseron.nvr.fronteast   ] [DEBUG   ] - NVR thread initialized
[2021-05-28 11:03:55] [viseron.nvr.frontwest   ] [DEBUG   ] - Waiting for first frame
[2021-05-28 11:03:55] [viseron.nvr.backeast    ] [DEBUG   ] - Waiting for first frame
[2021-05-28 11:03:55] [viseron.nvr.fronteast   ] [DEBUG   ] - Waiting for first frame
[2021-05-28 11:03:55] [viseron                 ] [INFO    ] - Initialization complete
[2021-05-28 11:03:55] [viseron.mqtt            ] [DEBUG   ] - MQTT connected with returncode 0
[2021-05-28 11:03:55] [viseron.nvr.frontwest   ] [DEBUG   ] - First frame received
[2021-05-28 11:03:56] [viseron.nvr.fronteast   ] [DEBUG   ] - First frame received
[2021-05-28 11:03:56] [viseron.recorder.fronteast] [INFO    ] - Starting recorder
[2021-05-28 11:03:56] [viseron.cleanup         ] [DEBUG   ] - Pausing segment cleanup
[2021-05-28 11:03:56] [viseron.nvr.fronteast   ] [INFO    ] - Starting motion detector
Exception in thread Thread-7:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/src/viseron/nvr.py", line 550, in run
    elif self.recorder.is_recording and self.event_over(processed_object_frame):
  File "/src/viseron/nvr.py", line 266, in event_over
    if self._object_filters[obj.label].require_motion:
KeyError: 'truck'
```

It's vastly different from the other error. SO I ran "docker exec -it viseron cat /detectors/models/darknet/coco.names" to make sure it is on the list and spelled correctly - and it is.

When I removed the "- label: truck" lines, it functions.

**Reply by gniknalu** [reply]
Having all kinds of issues with this version! (Haven't had a 'detection' since I started using it.)  Reverting to the older one.

**Reply by roflcoopter** [reply]
Sorry to hear that, I will look into it

**Reply by roflcoopter** [reply]
Just checking in on this again, is this still a problem in the dev tag?

**Reply by gniknalu** [reply]
> Just checking in on this again, is this still a problem in the dev tag?

I JUST tried it about 30 minutes ago. This is what I see in the logs:

```
[2021-06-01 15:44:38] [INFO    ] [viseron.recorder.fronteast] - Starting recorder
[2021-06-01 15:44:38] [INFO    ] [viseron.nvr.fronteast   ] - Starting motion detector
Exception in thread viseron.nvr.fronteast:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/src/viseron/nvr.py", line 553, in run
    elif self.recorder.is_recording and self.event_over(processed_object_frame):
  File "/src/viseron/nvr.py", line 271, in event_over
    if self._object_filters[obj.label].require_motion:
KeyError: 'car'
[2021-06-01 15:44:44] [DEBUG   ] [viseron.nvr.backeast    ] - First frame received
[2021-06-01 15:44:46] [DEBUG   ] [viseron.nvr.backeast.object] - Objects: [], message repeated 2 times
Exception in thread viseron.nvr.fronteast:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/src/viseron/nvr.py", line 553, in run
    elif self.recorder.is_recording and self.event_over(processed_object_frame):
  File "/src/viseron/nvr.py", line 271, in event_over
    if self._object_filters[obj.label].require_motion:
[2021-06-01 15:45:01] [DEBUG   ] [viseron.nvr.backeast.object] - Objects: [], message repeated 17 times
Exception in thread viseron.nvr.fronteast:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/src/viseron/nvr.py", line 553, in run
    elif self.recorder.is_recording and self.event_over(processed_object_frame):
  File "/src/viseron/nvr.py", line 271, in event_over
    if self._object_filters[obj.label].require_motion:
```

When I remove it, which it is located in two places, it works correctly.

This is one place in the config:

```
    zones:
      - name: EDrivewayRegion
        points:
          - x: 670
            y: 1073
          - x: 649
            y: 453
          - x: 926
            y: 422
          - x: 1200
            y: 196
          - x: 1913
            y: 253
          - x: 1913
            y: 498
          - x: 1906
            y: 1073
        labels:
          - label: person
            confidence: 0.8
            triggers_recording: true
          - label: car
            confidence: 0.8
            triggers_recording: true
            require_motion: true
```

**Reply by roflcoopter** [reply]
Fixed in latest dev.
It was caused by using `require_motion: true` on a label that wasnt being tracked under the global labels.
Can you try it out for me?

**Reply by gniknalu** [reply]
> Fixed in latest dev.
> It was caused by using `require_motion: true` on a label that wasnt being tracked under the global labels.
> Can you try it out for me?

I'll load it right now! Hold while I experiment. **Thank you!**

**Reply by gniknalu** [reply]
> Fixed in latest dev.
> It was caused by using `require_motion: true` on a label that wasnt being tracked under the global labels.
> Can you try it out for me?

No config errors with the latest dev version.

In the zone, there is a detection (car) and the zone turns green and it records. Then it repeats. And repeats. But the car is parked. 

Maybe my approach is incorrect? When a new car/truck pulls into the driveway I want a detection - different from the detection on the items that are parked in the driveway.

**Reply by roflcoopter** [reply]
`require_motion` only stops the recording, but not the detection. I guess this is a little bit unclear.
Am i understanding you right that you dont want the MQTT notifications either?

**Reply by gniknalu** [reply]
> `require_motion` only stops the recording, but not the detection. I guess this is a little bit unclear.
> Am i understanding you right that you dont want the MQTT notifications either?

I'm hoping to use the MQTT notification to trigger a picture sent via pushbullet. (Something I'm working on figuring out how to do using the MQTT payload.)

But when I add "require_motion" the goal is to only record when the detected object has motion. For example, two cars in the driveway - detect, fine but don't save a snapshot/video. Another car pulls into the driveway - detect yes but also record (as it is in motion or is a new detection).

**Reply by roflcoopter** [reply]
Ahh now i see the issue.
The require_motion flag only affects stopping of a recording, not starting of it.
Will fix and let you know when its up

Sorry you're having so many issues!

**Reply by gniknalu** [reply]
> Sorry you're having so many issues!

I'm enjoying the journey.  
No worries.

**Comment by roflcoopter** [comment]
I think i got what you need now in latest `dev`

I did some testing, walked infront of my camera so it started recording and then stood still, the recorder stopped when there was no motion and it did not restart until i moved. 
This solves the issue you were having, yes?

**Reply by gniknalu** [reply]
> I think i got what you need now in latest `dev`
> 
> I did some testing, walked infront of my camera so it started recording and then stood still, the recorder stopped when there was no motion and it did not restart until i moved.
> This solves the issue you were having, yes?

I'll update right now and will reply with the results. **Thank you!**

**Reply by gniknalu** [reply]
> I think i got what you need now in latest `dev`
> This solves the issue you were having, yes?

Updated. It is no longer recording on the items marked "require_motion" although they're detected. 
**BUT** it is no longer recording on **any** detections. Here is an example. I walked in front of two cameras. They both report something similar:

```
[2021-06-04 08:41:13] [DEBUG   ] [viseron.nvr.fronteast.object] - Objects: [{'label': 'person', 'confidence': 0.997, 'rel_width': 0.151, 'rel_height': 0.528, 'rel_x1': 0.604, 'rel_y1': 0.35, 'rel_x2': 0.755, 'rel_y2': 0.878}]
[2021-06-04 08:41:18] [DEBUG   ] [viseron.nvr.frontwest.object] - Objects: [{'label': 'person', 'confidence': 0.995, 'rel_width': 0.117, 'rel_height': 0.52, 'rel_x1': 0.334, 'rel_y1': 0.181, 'rel_x2': 0.451, 'rel_y2': 0.701}]

```
Both clearly detected me but neither recorded.

For now I'll be everting to the "latest" and commenting out the "require_motion" lines.

**Reply by roflcoopter** [reply]
😢
Can you show your full config when you ran this test?

**Reply by gniknalu** [reply]
> 😢
> Can you show your full config when you ran this test?

```
cameras:
  - name: "BackEast"
    host: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    port: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    stream_format: rtsp
    username: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    password: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    path: /.................................
    mqtt_name: camera_back_east
    publish_image: true
    motion_detection:
      interval: 1
      trigger_detector: false
      threshold: 15
      alpha: 0.6
      max_timeout: 300
      mask:
        - points:
            - x: 1914
              y: 1
            - x: 320
              y: 1
            - x: 320
              y: 148
            - x: 872
              y: 33
            - x: 1914
              y: 174
    object_detection:
      interval: 1

  - name: "FrontWest"
    host: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    port: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    stream_format: rtsp
    username: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    password: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    path: /.................................
    mqtt_name: camera_front_west
    publish_image: true
    zones:
      - name: DrivewayRegion
        points:
          - x: 213
            y: 158
          - x: 765
            y: 83
          - x: 1136
            y: 121
          - x: 1264
            y: 257
          - x: 1192
            y: 445
          - x: 1274
            y: 478
          - x: 1274
            y: 543
          - x: 608
            y: 713
          - x: 384
            y: 712
          - x: 204
            y: 368
        labels:
          - label: person
            confidence: 0.8
            triggers_recording: true
          - label: dog
            confidence: 0.8
            triggers_recording: true
          - label: cat
            confidence: 0.8
            triggers_recording: true
          - label: truck
            confidence: 0.8
            triggers_recording: true
            require_motion: true
          - label: car
            confidence: 0.8
            triggers_recording: true
            require_motion: true
    motion_detection:
      interval: .5
      trigger_detector: false
      timeout: true
      max_timeout: 330
      mask:
        - points:
            - x: 153
              y: 1
            - x: 1276
              y: 1
            - x: 1276
              y: 100
            - x: 748
              y: 65
            - x: 209
              y: 105
    object_detection:
      interval: .5
  - name: "FrontEast"
    host: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    port: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    stream_format: rtsp
    username: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    password: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    path: /.................................
    mqtt_name: camera_front_east
    publish_image: true
    zones:
      - name: EDrivewayRegion
        points:
          - x: 670
            y: 1073
          - x: 649
            y: 453
          - x: 926
            y: 422
          - x: 1200
            y: 196
          - x: 1913
            y: 253
          - x: 1913
            y: 498
          - x: 1906
            y: 1073
        labels:
          - label: person
            confidence: 0.8
            triggers_recording: true
          - label: dog
            confidence: 0.8
            triggers_recording: true
          - label: cat
            confidence: 0.8
            triggers_recording: true
          - label: truck
            confidence: 0.8
            triggers_recording: true
            require_motion: true
          - label: car
            confidence: 0.8
            triggers_recording: true
            require_motion: true
    motion_detection:
      interval: .5
      trigger_detector: false
      timeout: true
      max_timeout: 330
      mask:
        - points:
            - x: 1
              y: 1
            - x: 1912
              y: 1
            - x: 1910
              y: 237
            - x: 1609
              y: 196
            - x: 1404
              y: 126
            - x: 1123
              y: 122
            - x: 922
              y: 416
            - x: 1
              y: 455
    object_detection:
      interval: .5
      
  - name: "BackWest"
    host: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    port: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    stream_format: rtsp
    username: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    password: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
    path: /.................................
    mqtt_name: camera_back_west
    publish_image: true
    motion_detection:
      interval: 1
      trigger_detector: true
      timeout: true
      max_timeout: 300
    object_detection:
      interval: .5
    logging:
      level: debug

recorder:
  extension: mkv
    filename_pattern: '%Y-%m-%d-%I%M%S'
  lookback: 10
  timeout: 10
  retain: 30
  thumbnail:
    filename_pattern: '%Y-%m-%d-%I%M%S'

# MQTT is optional
mqtt:
 broker: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
 port: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
 username: !secret xxxxxxxxxxxxxxxxxxxxxxxxx
 password: !secret xxxxxxxxxxxxxxxxxxxxxxxxx

logging:
  level: debug
```

**Reply by roflcoopter** [reply]
I am not seeing this behaviour.
Running with `require_motion: false` and `require_motion: true`  both triggers the recorder for me.

Could you post the full log from Viseron?

**Reply by gniknalu** [reply]
> Could you post the full log from Viseron?

3 of the 4 cameras above were enabled. I grabbed the latest dev image and started it. Then I walked in front of each of the 3 cameras. Verified a "detection" took place as it said "Person" on each camera. This is reflected in the logs. Two of the cameras already had a "green" zone with a detection of "truck" but there is no evidence in the logs of this happening. Then, afterwards, I checked to see if any video/snapshot was written. No files were generated. 

Stopped and rm the dev image. Replaced the conf file with one that supports the "latest" image. Started the "latest" image. Repeated the test above and it did write video/snapshot files.

[Here is where you can find the full log of what happened above](https://gist.github.com/gniknalu/831986c751bbff9458dc8e1ec00a680d)

**Comment by roflcoopter** [comment, marked as answer]
I made a silly mistake...

It should work in latest dev (have you heard this one before?) 😂

It takes about 10 minutes to build so you can pull and test in a bit.

**Reply by gniknalu** [reply]
> I made a silly mistake...
> 
> It should work in latest dev (have you heard this one before?) 😂
> 
> It takes about 10 minutes to build so you can pull and test in a bit.

It's better than testing in production! 😂

I'll grab it and test in a few.

Thank you again!

**Reply by roflcoopter** [reply]
Yeah... I need to start writing unittests but that is so much work

**Reply by gniknalu** [reply]
> Yeah... I need to start writing unittests but that is so much work

One thing I do in my 'day job' is automated regression testing. We should talk.

So far it's working as expected. I'll let it run through the day and will report back.

**Thank you for your excellent product!**

**Reply by gniknalu** [reply]
> I made a silly mistake...
> 
> It should work in latest dev (have you heard this one before?) 😂
> 
> It takes about 10 minutes to build so you can pull and test in a bit.

Ran some tests. With the truck parked in the driveway - no recording. (Exactly as I would like.) When I went to the truck - the recording was triggered on detecting a "person".  (As Expected.) But when I arrive, it only starts recording when I get out of the truck but doesn't "detect" and record when the truck pulls in. Well it DOES detect (I can see it turn green and say "truck") but doesn't record. 

I'll run more definitive tests tomorrow especially since we'll have company pulling into the driveway. 

Thank you again.

**Reply by roflcoopter** [reply]
> > Yeah... I need to start writing unittests but that is so much work
> 
> One thing I do in my 'day job' is automated regression testing. We should talk.
> 
> So far it's working as expected. I'll let it run through the day and will report back.
> 
> **Thank you for your excellent product!**

That sounds amazing! What tools are you using to do that?

**Reply by gniknalu** [reply]
> That sounds amazing! What tools are you using to do that?

I'm using jMeter - currently for some .Net apps. I'm sure it could be adapted - somehow - for this app. (Possibly adding 'supportability' into this app for improved regression testing?)

**Reply by roflcoopter** [reply]
Okay i see, googled a little, is jMeter used to call APIs only?

**Reply by gniknalu** [reply]
> Okay i see, googled a little, is jMeter used to call APIs only?

I use jMeter for webwalks, REST, SOAP etc. There are plugins for all types of testing. 
Maybe if there was a way to test via MQTT using jMeter to listen for certain messages? Would need to publish more information though for it to be useful.
